### PR TITLE
implement gemm and memory_cache

### PR DESCRIPTION
**Pull request by okdshin**
`memory_cache` is the base building block for mkldnn backend.
It has multiple memories which may have different dims or memory format from each other but they share data contents.

### CODE REVIEW

File: menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp [hunk 1]
```diff
@@ -39,61 +39,81 @@ namespace menoh_impl {
...
+                                                   "parameter table"
+                                                << std::endl;
+                                        auto result_pair =
+                                          variable_memory_cache_table_.emplace(
+                                            input_name,
+                                            memory_cache(found->second,
+                                                         engine_));
+                                        assert(
+                                          result_pair.second &&
+                                          "alredy same named variable exist");
```
[the remark this answers is not in the document]
Comment (okdshin) [reply menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp]: Thanks. I'll fix it.

File: menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp [hunk 2]
```diff
@@ -113,23 +133,55 @@ namespace menoh_impl {
...
                                           copy_proc);
-                                        input_list.push_back(arr);
-                                        is_found_from_other_context = true;
+                                        auto result_pair =
+                                          variable_memory_cache_table_.emplace(
+                                            input_name,
+                                            memory_cache(arr, engine_));
+                                        assert(
+                                          result_pair.second &&
+                                          "alredy same named variable exist");
```
Comment (msakai) [inline menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp]:
```suggestion
                                          "already same named variable exist");
```

File: menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (msakai) [reply menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp]: Also, does taking `&&` of `bool` and string literal make sense as an argument of `assert`?
I guess you intended to throw an exception.

[the remark this answers is not in the document]
Comment (okdshin) [reply menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp]: It's my failure. I'll fix.

[the remark this answers is not in the document]
Comment (okdshin) [reply menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp]: No. It is actually right. `result_pair.second` will true if emplacing is succeeded. A string literal is non null `const char*`. So this code works as I intented.
`assert(cond && "string literal")` is a famous technique introduced in "Herb Sutter, Andrei Alexandrescu. C++ coding standards: 101 rules, guidelines, and best practices. ISBN: 0-32-111358-6. Chapter 68: 'Assert liberally to document internal assumptions and invariants'"
It also introduces `assert(!"string literal")`

[the remark this answers is not in the document]
Comment (msakai) [reply menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp]: I see.
That's an interesting hack!

**Comment by okdshin** [comment]
Thanks review. By using valgrind, I found this code (and other PRs fo other operators) have double free corruption code. I will modify them.

**Comment by okdshin** [comment]
I integrated this PR and PRs for other operators and tried to infer VGG16. I found this PR has some issues so I'll fix them. Please review then.

**Comment by okdshin** [comment]
I modified. Please review! @msakai

**Comment by msakai** [comment]
Sure.

File: menoh/mkldnn_with_generic_fallback/backend/mkldnn/operator/gemm.hpp [hunk 3]
```diff
@@ -98,6 +94,7 @@ namespace menoh_impl {
                   gemm_output_md);
                 auto gemm_pd = mkldnn::inner_product_forward::primitive_desc(
                   gemm_desc, engine);
+                auto op_input_dims = extract_dims(gemm_pd.src_primitive_desc());
```
Comment (msakai) [inline menoh/mkldnn_with_generic_fallback/backend/mkldnn/operator/gemm.hpp]: `op_input_dims` is unused. Is it intentional?

**Comment by msakai** [comment]
> Thanks review. By using valgrind, I found this code (and other PRs fo other operators) have double free corruption code. I will modify them.

Where was the "double free corruption code" and which commit fixed the problem?

File: menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp [hunk 4]
```diff
@@ -161,9 +229,18 @@ namespace menoh_impl {
...
                       .submit(primitive_list)
                       .wait();
+                    */
+                    int i = 0;
+                    for(auto const& primitive : primitive_list) {
+                        mkldnn::stream(mkldnn::stream::kind::eager)
+                          .submit({primitive})
+                          .wait();
+                        ++i;
+                    }
```
[the remark this answers is not in the document]
Comment (okdshin) [reply menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp]: Interupting stream execution does not affect the speed directly. https://github.com/intel/mkl-dnn/issues/328
However, there may be some overheads. I'll fix it.

I forgot to remove `i`. I'll fix it as well.

File: menoh/mkldnn_with_generic_fallback/backend/mkldnn/operator/gemm.hpp (hunk 3 above)
Comment (okdshin) [reply menoh/mkldnn_with_generic_fallback/backend/mkldnn/operator/gemm.hpp]: No. It was for debugging. I'll remove it.

**Comment by okdshin** [comment]
Actually I found two bugs. One is double free corruption [here](https://github.com/pfnet-research/menoh/pull/137/commits/08761182e11e07c31565cde9e5f1f71b90d6763f#diff-791f7ab18a103d6167001631aec9ec24L122). Both `op_output_memory` and `output_memory` free the shared buffer. To fix it, I stopped to use `mkldnn::memory` for passing output_memory. I added `formatted_array` class to replace it and `manage_output` function to gather common lines between operators.
Another bug is a conventional `invalidated iterator` bug. [here](https://github.com/pfnet-research/menoh/pull/137/commits/e0d74fbe29cf9eb9b131710ea5d85477a3d162ac#diff-edb22fc899793fed1e5174786589ce57L60), `add_cached_memory(new_memory)` caused vector's reallocation and invalidate `found_memory` const reference [here](https://github.com/pfnet-research/menoh/pull/137/commits/e0d74fbe29cf9eb9b131710ea5d85477a3d162ac#diff-edb22fc899793fed1e5174786589ce57L55).

**Comment by msakai** [comment]
Thank you for the explanation, but I'm a bit confused.

https://github.com/pfnet-research/menoh/pull/137/commits/08761182e11e07c31565cde9e5f1f71b90d6763f#diff-791f7ab18a103d6167001631aec9ec24L122 :
```cxx
                    op_output_memory = mkldnn::memory(
                      {{{extract_dims(output_memory)},
                        extract_data_type(output_memory),
                        extract_format(gemm_pd.dst_primitive_desc())},
                       engine},
                      output_memory.get_data_handle());
                    output_memory_cache.add_cached_memory(*op_output_memory);
```

Here, [`memory(const primitive_desc &adesc, void *ahandle)`](https://github.com/intel/mkl-dnn/blob/4e333787e0d66a1dca1218e99a891d493dbc8ef1/include/mkldnn.hpp#L766-L773) is called. 

Unlike [`memory(const primitive_desc &adesc)`](https://github.com/intel/mkl-dnn/blob/4e333787e0d66a1dca1218e99a891d493dbc8ef1/include/mkldnn.hpp#L739-L764), this constructor does not `reset` the `_handle` field. Therefore it seems that `op_output_memory` does not free the shared buffer?

**Comment by msakai** [comment]
Anyway, my confusion is about the past code and my concern might be irrelevant to the current code.

Current code itself looks good to me.

**Comment by okdshin** [comment]
You are correct. I misunderstood that. [here](https://github.com/pfnet-research/menoh/commit/08761182e11e07c31565cde9e5f1f71b90d6763f#diff-6f96fe34d5fa00be3313c1bceffe0db9L170) seems the actual problem line. In this line, array is not stored so buffer is invalidated. (I can't understsand why that bug causes double free corruption tho). Anyway thank you for pointing out.

**Comment by okdshin** [comment]
Thank you for review @msakai !

**Pull request merged**